File: clients/haskell/src/Model/ActionType.hs
module Model.ActionType where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int

-- | Type of action a unit is currently performing
data ActionType
    -- | Picking up or dropping loot
    = Looting
    -- | Using a shield potion
    | UseShieldPotion
    deriving (Eq, Ord, Show)

instance Trans ActionType where
    read = do
        tag :: Int32 <- Trans.read
        return $ case tag of
            0 -> Looting
            1 -> UseShieldPotion
            
    write Looting =
        Trans.write (0 :: Int32)
    write UseShieldPotion =
        Trans.write (1 :: Int32)

File: clients/swift/Sources/AiCup22/Model/ActionType.swift
/// Type of action a unit is currently performing
public enum ActionType: Int32 {
    /// Picking up or dropping loot
    case looting = 0

    /// Using a shield potion
    case useShieldPotion = 1

    /// Read ActionType from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> ActionType {
        return ActionType(rawValue: stream.readInt32())!
    }

    /// Write ActionType to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(rawValue)
    }
}
